File: windows/tests/SkillHub.Core.Tests/ScannerTests.cs
namespace SkillHub.Core.Tests;

sealed class TemporaryHome : IDisposable
{
    public string Url { get; }
    public HubPaths Paths { get; }

    public TemporaryHome()
    {
        Url = Path.Combine(Path.GetTempPath(), $"skillhub-tests-{Guid.NewGuid():N}");
        Directory.CreateDirectory(Url);
        Paths = new HubPaths(Url);
    }

    public string MakeSkill(ToolSource source, string folder, string? name = null, string description = "desc")
    {
        var root = Paths.SkillRoot(source) ?? throw new InvalidOperationException();
        var dir = Path.Combine(root, folder);
        Directory.CreateDirectory(dir);
        PathUtil.WriteAtomic(Path.Combine(dir, "SKILL.md"), Frontmatter.Render(name ?? folder, description, $"# {folder}\n"));
        return dir;
    }

    public void Link(ToolSource source, string folder, string target)
    {
        var root = Paths.SkillRoot(source) ?? throw new InvalidOperationException();
        Directory.CreateDirectory(root);
        Directory.CreateSymbolicLink(Path.Combine(root, folder), target);
    }

    public void Dispose()
    {
        try
        {
            Directory.Delete(Url, true);
        }
        catch (IOException)
        {
        }
    }
}

public class ScannerTests
{
    [Fact]
    public void DedupesSymlinkedInstallsIntoOneSkill()
    {
        using var home = new TemporaryHome();
        var real = home.MakeSkill(ToolSource.Agents, "blog-write");
        home.Link(ToolSource.CursorUser, "blog-write", real);
        home.Link(ToolSource.Claude, "blog-write", real);

        var snapshot = Scanner.Scan(home.Paths, new ScanOptions());
        Assert.Single(snapshot.Skills);
        var skill = snapshot.Skills[0];
        Assert.Equal(3, skill.Installations.Count);
        Assert.Equal(2, skill.SymlinkCount);
        Assert.Equal(1, skill.RealCount);
        Assert.True(skill.IsDuplicateInstall);
        Assert.Empty(skill.Health);
        Assert.Equal(
            new HashSet<ToolSource> { ToolSource.Agents, ToolSource.CursorUser, ToolSource.Claude },
            skill.ToolSources.ToHashSet());
    }

    [Fact]
    public void ReportsBrokenLinksAndMissingDescriptions()
    {
        using var home = new TemporaryHome();
        home.MakeSkill(ToolSource.Claude, "quiet", description: "");
        home.Link(ToolSource.Codex, "gone", Path.Combine(home.Url, "nowhere"));

        var snapshot = Scanner.Scan(home.Paths, new ScanOptions());
        Assert.Single(snapshot.Skills);
        Assert.Equal([HealthIssue.MissingDescription], snapshot.Skills[0].Health);
        Assert.Single(snapshot.BrokenOrphans);
        Assert.Equal(ToolSource.Codex, snapshot.BrokenOrphans[0].Source);

        var stats = OverviewStats.From(snapshot);
        Assert.Equal(1, stats.MissingDescriptions);
        Assert.Equal(1, stats.BrokenLinks);
    }

    [Fact]
    public void BuiltinSkillsAreReadOnly()
    {
        using var home = new TemporaryHome();
        home.MakeSkill(ToolSource.CursorBuiltin, "builtin");
        home.MakeSkill(ToolSource.CursorUser, "mine");

        var snapshot = Scanner.Scan(home.Paths, new ScanOptions());
        Assert.True(snapshot.Skills.First(s => s.Name == "builtin").IsReadOnly);
        Assert.False(snapshot.Skills.First(s => s.Name == "mine").IsReadOnly);
    }

    [Fact]
    public void FindsEmbeddedAndStandalonePrompts()
    {
        using var home = new TemporaryHome();
        var skill = home.MakeSkill(ToolSource.Claude, "writer");
        var promptsDir = Path.Combine(skill, "prompts");
        Directory.CreateDirectory(promptsDir);
        PathUtil.WriteAtomic(Path.Combine(promptsDir, "outline.md"), "# Outline");
        PathUtil.WriteAtomic(Path.Combine(skill, "README.md"), "ignored");
        home.Paths.EnsureContentDirectories();
        PathUtil.WriteAtomic(Path.Combine(home.Paths.PromptLibrary, "weekly-report.md"), "# Standalone");

        var snapshot = Scanner.Scan(home.Paths, new ScanOptions());
        Assert.Equal(2, snapshot.Prompts.Count);
        var embedded = snapshot.Prompts.First(p => p.Kind == PromptKind.Embedded);
        Assert.Equal("outline", embedded.Title);
        Assert.Equal("writer", embedded.ParentSkillName);
        var standalone = snapshot.Prompts.First(p => p.Kind == PromptKind.Standalone);
        Assert.Equal("weekly report", standalone.Title);
        Assert.Equal(ToolSource.PromptLibrary, standalone.Source);
    }

    [Fact]
    public void ProjectSkillsOnlyWhenEnabled()
    {
        using var home = new TemporaryHome();
        var project = Path.Combine(home.Url, "Projects", "demo", ".cursor", "skills", "local-skill");
        Directory.CreateDirectory(project);
        PathUtil.WriteAtomic(Path.Combine(project, "SKILL.md"), Frontmatter.Render("local-skill", "d", ""));

        Assert.Empty(Scanner.Scan(home.Paths, new ScanOptions()).Skills);
        var enabled = Scanner.Scan(home.Paths, new ScanOptions(ScanProjectSkills: true));
        Assert.Single(enabled.Skills);
        Assert.Equal(ToolSource.Custom, enabled.Skills[0].Installations[0].Source);
    }

    [Fact]
    public void WritableRootsExcludeUnrelatedPaths()
    {
        using var home = new TemporaryHome();
        var roots = home.Paths.WritableRoots(new ScanOptions());
        Assert.Contains(PathUtil.Standardize(home.Paths.ContentRoot), roots);
        Assert.Contains(PathUtil.Standardize(home.Paths.SkillRoot(ToolSource.Claude)!), roots);
        Assert.DoesNotContain(PathUtil.Standardize(home.Url), roots);
    }
}

File: SkillHubTests/ScannerTests.swift
import Foundation
import Testing
@testable import SkillHub

/// Builds a throwaway home directory that mimics the real tool folders.
struct TemporaryHome {
    let url: URL
    let paths: HubPaths

    init() throws {
        url = FileManager.default.temporaryDirectory
            .appending(path: "skillhub-tests-\(UUID().uuidString)", directoryHint: .isDirectory)
        try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        paths = HubPaths(home: url)
    }

    func makeSkill(_ source: ToolSource, folder: String, name: String? = nil, description: String = "desc") throws -> URL {
        let root = try #require(paths.skillRoot(source))
        let dir = root.appending(path: folder, directoryHint: .isDirectory)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let text = Frontmatter.render(name: name ?? folder, description: description, body: "# \(folder)\n")
        try text.write(to: dir.appending(path: "SKILL.md"), atomically: true, encoding: .utf8)
        return dir
    }

    func link(_ source: ToolSource, folder: String, to target: URL) throws {
        let root = try #require(paths.skillRoot(source))
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        try FileManager.default.createSymbolicLink(at: root.appending(path: folder), withDestinationURL: target)
    }

    func tearDown() {
        try? FileManager.default.removeItem(at: url)
    }
}

struct ScannerTests {
    @Test func dedupesSymlinkedInstallsIntoOneSkill() throws {
        let home = try TemporaryHome()
        defer { home.tearDown() }

        let real = try home.makeSkill(.agents, folder: "blog-write")
        try home.link(.cursorUser, folder: "blog-write", to: real)
        try home.link(.claude, folder: "blog-write", to: real)

        let snapshot = Scanner.scanSync(paths: home.paths, options: ScanOptions())
        #expect(snapshot.skills.count == 1)
        let skill = try #require(snapshot.skills.first)
        #expect(skill.installations.count == 3)
        #expect(skill.symlinkCount == 2)
        #expect(skill.realCount == 1)
        #expect(skill.isDuplicateInstall)
        #expect(skill.health.isEmpty)
        #expect(Set(skill.toolSources) == [.agents, .cursorUser, .claude])
    }

    @Test func reportsBrokenLinksAndMissingDescriptions() throws {
        let home = try TemporaryHome()
        defer { home.tearDown() }

        _ = try home.makeSkill(.claude, folder: "quiet", description: "")
        try home.link(.codex, folder: "gone", to: home.url.appending(path: "nowhere"))

        let snapshot = Scanner.scanSync(paths: home.paths, options: ScanOptions())
        #expect(snapshot.skills.count == 1)
        #expect(snapshot.skills.first?.health == [.missingDescription])
        #expect(snapshot.brokenOrphans.count == 1)
        #expect(snapshot.brokenOrphans.first?.source == .codex)

        let stats = OverviewStats(snapshot: snapshot)
        #expect(stats.missingDescriptions == 1)
        #expect(stats.brokenLinks == 1)
    }

    @Test func builtinSkillsAreReadOnly() throws {
        let home = try TemporaryHome()
        defer { home.tearDown() }

        _ = try home.makeSkill(.cursorBuiltin, folder: "builtin")
        _ = try home.makeSkill(.cursorUser, folder: "mine")

        let snapshot = Scanner.scanSync(paths: home.paths, options: ScanOptions())
        let builtin = try #require(snapshot.skills.first { $0.name == "builtin" })
        let mine = try #require(snapshot.skills.first { $0.name == "mine" })
        #expect(builtin.isReadOnly)
        #expect(!mine.isReadOnly)
    }

    @Test func findsEmbeddedAndStandalonePrompts() throws {
        let home = try TemporaryHome()
        defer { home.tearDown() }

        let skill = try home.makeSkill(.claude, folder: "writer")
        let promptsDir = skill.appending(path: "prompts", directoryHint: .isDirectory)
        try FileManager.default.createDirectory(at: promptsDir, withIntermediateDirectories: true)
        try "# Outline".write(to: promptsDir.appending(path: "outline.md"), atomically: true, encoding: .utf8)
        try "ignored".write(to: skill.appending(path: "README.md"), atomically: true, encoding: .utf8)

        try home.paths.ensureContentDirectories()
        try "# Standalone".write(to: home.paths.promptLibrary.appending(path: "weekly-report.md"), atomically: true, encoding: .utf8)

        let snapshot = Scanner.scanSync(paths: home.paths, options: ScanOptions())
        #expect(snapshot.prompts.count == 2)
        let embedded = try #require(snapshot.prompts.first { $0.kind == .embedded })
        #expect(embedded.title == "outline")
        #expect(embedded.parentSkillName == "writer")
        let standalone = try #require(snapshot.prompts.first { $0.kind == .standalone })
        #expect(standalone.title == "weekly report")
        #expect(standalone.source == .promptLibrary)
    }

    @Test func projectSkillsOnlyWhenEnabled() throws {
        let home = try TemporaryHome()
        defer { home.tearDown() }

        let project = home.url.appending(path: "Projects/demo/.cursor/skills/local-skill", directoryHint: .isDirectory)
        try FileManager.default.createDirectory(at: project, withIntermediateDirectories: true)
        try Frontmatter.render(name: "local-skill", description: "d", body: "")
            .write(to: project.appending(path: "SKILL.md"), atomically: true, encoding: .utf8)

        #expect(Scanner.scanSync(paths: home.paths, options: ScanOptions()).skills.isEmpty)
        let enabled = Scanner.scanSync(paths: home.paths, options: ScanOptions(scanProjectSkills: true))
        #expect(enabled.skills.count == 1)
        #expect(enabled.skills.first?.installations.first?.source == .custom)
    }

    @Test func writableRootsExcludeUnrelatedPaths() throws {
        let home = try TemporaryHome()
        defer { home.tearDown() }
        let roots = home.paths.writableRoots(options: ScanOptions())
        #expect(roots.contains(home.paths.contentRoot.path))
        #expect(roots.contains(home.paths.skillRoot(.claude)!.path))
        #expect(!roots.contains(home.url.path))
    }
}
